app/src/main/java/edu/wpi/first/shuffleboard/app/plugin/PluginLoader.java

```diff
@@ -0,0 +1,189 @@
+   */
+  public void loadPluginJar(URI jarUri) throws IOException {
+    URL url = jarUri.toURL();
+    PrivilegedAction<URLClassLoader> getClassLoader = () -> {
+      return new URLClassLoader(new URL[]{url}, ClassLoader.getSystemClassLoader());
+    };
+    URLClassLoader classLoader = AccessController.doPrivileged(getClassLoader);
+    try (JarFile jarFile = new JarFile(new File(jarUri))) {
+      jarFile.stream()
+          .filter(e -> e.getName().endsWith(".class"))
+          .map(e -> e.getName().replace('/', '.'))
+          .map(n -> n.substring(0, n.length() - 6)) // ".class".length() == 6
+          .flatMap(className -> tryLoadClass(className, classLoader))
+          .forEach(this::loadPluginClass);
+    }
+  }
+
+  private Stream<Class<?>> tryLoadClass(String name, ClassLoader classLoader) {
+    try {
+      return Stream.of(Class.forName(name, false, classLoader));
+    } catch (ClassNotFoundException e) {
+      // TODO log
+      return Stream.empty();
+    }
+  }
+
+  /**
+   * Attempts to load a plugin class. This class may or not be a plugin; only a plugin class will be loaded.
+   *
+   * @param clazz the class to attempt to load
+   *
+   * @return true if the class is a plugin class and was successfully loaded; false otherwise
+   */
+  public boolean loadPluginClass(Class<?> clazz) {
+    if (Plugin.class.isAssignableFrom(clazz) && !Modifier.isAbstract(clazz.getModifiers())) {
+      try {
+        if (Modifier.isPublic(clazz.getConstructor().getModifiers())) {
+          Plugin plugin = (Plugin) clazz.newInstance();
+          // Unload existing plugin, if it exists
+          loadedPlugins.stream()
+              .filter(p -> p.idString().equals(plugin.idString()))
+              .findFirst()
+              .ifPresent(oldPlugin -> {
+                unload(oldPlugin);
+                knownPlugins.remove(oldPlugin);
+              });
+          load(plugin);
+          return true;
+        }
+      } catch (ReflectiveOperationException e) {
+        log.log(Level.WARNING, "Could not load plugin class", e);
+        return false;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Loads a plugin.
+   *
+   * @param plugin the plugin to load
+   *
+   * @throws IllegalArgumentException if a plugin has already been loaded with the same name
+   */
+  public void load(Plugin plugin) {
+    if (loadedPlugins.contains(plugin)) {
+      throw new IllegalArgumentException("The plugin " + plugin + " is already loaded");
+    }
+    log.info("Loading plugin " + plugin.fullIdString()); //NOPMD log not in if statement
+    plugin.getDataTypes().forEach(DataTypes.getDefault()::register);
+    plugin.getSourceTypes().forEach(SourceTypes.getDefault()::register);
+    plugin.getTypeAdapters().forEach(Serializers::add);
+    plugin.getWidgets().forEach(Widgets.getDefault()::register);
+    plugin.getDefaultWidgets().forEach(Widgets.getDefault()::setDefaultWidget);
```

Comment:
Are you concerned that these are all using the static singleton of the `DataTypes`, `SourceTypes` ect.. objects?

I think the solution is to pas them as dependencies into the object and apply the dependency inversion principle here.